base/polly.jl

```diff
@@ -0,0 +1,152 @@
+immutable UnitRange{T<:Real} <: AbstractUnitRange{T}
+    start::T
+    stop::T
+    UnitRange(start, stop) = new(start, stop)
+end
+UnitRange{T<:Real}(start::T, stop::T) = Polly.UnitRange{T}(start, stop)
+
+# This method was directly adopted from `Base.UnitRange`.
+start{T}(r::Polly.UnitRange{T}) = oftype(r.start + one(T), r.start)
+
+# This has to be different than for `Base.UnitRange` to reflect the different
+# behavior of the `Polly.UnitRange` constructor.
+done{T}(r::Polly.UnitRange{T}, i) = (i < oftype(i, r.start)) | (i > oftype(i, r.stop))
+
+# `Base.StepRange` uses the same representation of emptiness as described above
+# for `Base.UnitRange` with `stop == start - 1` but additionally, in the case of
+# non-emptiness, its constructor will precompute the last value that is actually
+# part of the range. For example, `5:2:8` would actually result in `5:2:7`. In
+# `Polly.StepRange` we simplify construction by dropping these requirements,
+# i.e. `5:2:8` would also be used as is. When iterating over a `Polly.StepRange`
+# loop bounds will therefore be more obvious to Polly than with a
+# `Base.StepRange`.
+immutable StepRange{T,S} <: OrdinalRange{T,S}
+    start::T
+    step::S
+    stop::T
+    StepRange(start::T, step::S, stop::T) = new(start, step, stop)
+end
+StepRange{T,S}(start::T, step::S, stop::T) = Polly.StepRange{T,S}(start, step, stop)
+
+# This method was directly adopted from `Base.StepRange`.
+start(r::Polly.StepRange) = oftype(r.start + r.step, r.start)
+
+# This method was directly adopted from `Base.StepRange`.
+next{T}(r::Polly.StepRange{T}, i) = (convert(T,i), i + r.step)
+
+# We have to use a simpler condition as for `Base.StepRange` in order to
+# be able to derive the loop bounds in Polly. For now it also ignores
+# wrap-arounds which could for example happen for `1:1:typemax(Int64)` which we
+# consider a rare application.
+done{T,S}(r::Polly.StepRange{T,S}, i) = (r.step > zero(r.step)) ? (i > oftype(i, r.stop)) :
+                                                                  (i < oftype(i, r.stop))
+
+# This was directly adopted from `Base.UnitRange` and `Base.StepRange` to avoid
+# overflows for smaller `Integer` types.
+let smallint = (Int === Int64 ?
+                Union{Int8,UInt8,Int16,UInt16,Int32,UInt32} :
+                Union{Int8,UInt8,Int16,UInt16})
+    global start
+    global next
+    start{T<:smallint}(r::Polly.StepRange{T}) = convert(Int, r.start)
+    next{T<:smallint}(r::Polly.StepRange{T}, i) = (i % T, i + r.step)
+    start{T<:smallint}(r::Polly.UnitRange{T}) = convert(Int, r.start)
+end
+
+# Find assigments of the form `i = start:stop` and `i = start:step:stop` that
+# occur in `for`-loop headers in `func` and replace them by
+# `i = Polly.UnitRange(start,stop)` and `i = Polly.StepRange(start,step,stop)`.
+function canonicalize!(func)
+    worklist = [func]
+    while !isempty(worklist)
+        expr = pop!(worklist)
+        if expr.head == :for
+            loop_header = expr.args[1]
```

Comment:
Wow ! What's this feature of Julia that allows functions to be accessed like a parse tree ?